app/src/org/commcare/dalvik/activities/EntitySelectActivity.java

```diff
@@ -534,100 +496,102 @@ public void onItemClick(AdapterView<?> listView, View view, int position, long i
             i.putExtra("entity_detail_index", position);
             if (mNoDetailMode) {
                 returnWithResult(i);
-            } else  {
+            } else {
                 startActivityForResult(i, CONFIRM_SELECT);
             }
         }
     }
 
+    public void onBarcodeFetch(String result, Intent intent) {
+        setSearchText(result);
+    }
+
+    public void onCalloutResult(String result, Intent intent) {
+        boolean resultSet = false;
+        if (result != null) {
+            setSearchText(result);
+            resultSet = true;
+        }
+        Callout callout = shortSelect.getCallout();
+        for (String key : callout.getResponses()) {
+            result = intent.getExtras().getString(key);
+            if (result != null && !resultSet) {
+                resultSet = true;
+                setSearchText(result);
+                break;
+            }
+        }
+    }
+
     @Override
     protected void onActivityResult(int requestCode, int resultCode, Intent intent) {
-        switch(requestCode){
-        case BARCODE_FETCH:
-            if(resultCode == Activity.RESULT_OK) {
-                String result = intent.getStringExtra("SCAN_RESULT");
-                this.searchbox.setText(result);
-            }
-            break;
-        case CALLOUT:
-            if (resultCode == Activity.RESULT_OK) {
-                boolean resultSet = false;
-                String result = intent.getStringExtra("odk_intent_data");
-                if (result != null) {
-                    this.searchbox.setText(result);
-                    resultSet = true;
-                }
-                Callout callout = shortSelect.getCallout();
-                for (String key : callout.getResponses()) {
-                    result = intent.getExtras().getString(key);
-                    if (result != null && !resultSet) {
-                        resultSet = true;
-                        this.searchbox.setText(result);
-                        break;
+        switch (requestCode) {
+            case BarcodeScanListenerDefaultImpl.BARCODE_FETCH:
+                BarcodeScanListenerDefaultImpl.onBarcodeResult(this, requestCode, resultCode, intent);

```

Comment:
If we leave this extra layer of indirection, we can centralize error processing and reduce LoC. I'm planning on a bigger refactor for barcode-related stuff, so let's just leave it there for now.